Clean/Circuit/StructuralLemmas.lean

```diff
@@ -0,0 +1,255 @@
+      intro input offset
+      simp only [Circuit.bind_def, Circuit.output, circuit_norm]
+    subcircuitsConsistent := by
+      intro input offset
+      simp only [Circuit.bind_def, Circuit.operations, circuit_norm]
+      ring
+  }
+  Assumptions := circuit1.Assumptions
+  Spec := fun input output => ∃ mid, circuit1.Spec input mid ∧ circuit2.Spec mid output
+  soundness := by
+    apply Circuit.soundness_compose_circuits (Mid := Mid)
+    · intro; rfl
+    · intro input offset
+      simp only [ElaboratedCircuit.output]
+    · intro input h; exact h
+    · intro input mid h_assumptions h_spec1
+      exact h_compat input mid h_assumptions h_spec1
+    · intro input mid output h_input h_spec1 h_spec2
+      exact ⟨mid, h_spec1, h_spec2⟩
+  completeness := by
+    simp only [circuit_norm]
+    rintro offset env input_var ⟨ use1, use2 ⟩ input h_input asm1
+    constructor
+    · simp only [subcircuit_norm, h_input, asm1]
+    · simp only [subcircuit_norm]
+      apply h_compat
+      · apply asm1
+      · simp only [subcircuit_norm] at use1
+        simp only [h_input] at use1
+        apply use1
+        assumption
+}
+
+@[circuit_norm]
+lemma concat_Assumptions {F Input Mid Output} [Field F] [ProvableType Input] [ProvableType Mid] [ProvableType Output]
+    (c1 : FormalCircuit F Input Mid) (c2 : FormalCircuit F Mid Output) p0 p1 :
+    (c1.concat c2 p0 p1).Assumptions = c1.Assumptions := by
+  simp only [FormalCircuit.concat]
+
+/--
+Weaken the specification of a FormalCircuit.
+
+This combinator takes a FormalCircuit with a strong specification and produces
+a new FormalCircuit with a weaker specification. This is useful when:
+- You have a circuit that proves more than you need
+- You want to compose circuits where the specs don't match exactly
+- You need to adapt a specific circuit to a more general interface
+
+The requirements are:
+- The assumptions remain the same or can be strengthened
+- The stronger spec implies the weaker spec
+-/
+def FormalCircuit.weakenSpec
+    {F : Type} [Field F]
+    {Input Output : TypeMap} [ProvableType Input] [ProvableType Output]
+    (circuit : FormalCircuit F Input Output)
+    (WeakerSpec : Input F → Output F → Prop)
+    (h_spec_implication : ∀ input output,
+      circuit.Assumptions input →
+      circuit.Spec input output →
+      WeakerSpec input output) :
+    FormalCircuit F Input Output := {
+  elaborated := circuit.elaborated
+  Assumptions := circuit.Assumptions
+  Spec := WeakerSpec
+  soundness := by
+    intro offset env input_var input h_eval h_assumptions h_holds
+    -- Use the original circuit's soundness
+    have h_strong_spec := circuit.soundness offset env input_var input h_eval h_assumptions h_holds
+    -- Apply the implication to get the weaker spec
+    exact h_spec_implication input _ h_assumptions h_strong_spec
+  completeness := by
+    -- Completeness is preserved since we use the same elaborated circuit
+    -- and the same assumptions
+    exact circuit.completeness
+}
```

Comment:
Add a lemma called weakenSpec_Assumptions (similar to concat_Assumptions) and put that in the circuit_norm set.